lib/metrics.js

```diff
@@ -22,169 +14,56 @@ class MetricsEvent {
-      : this.allMetricsPath
-  }
-
-  /**
-   * @param {string=} module
-   * @returns {Promise<MetricsEvent>}
-   */
-  async getLatest (module) {
-    const metrics = await fs.readFile(this.#getMetricsFilePath(module), 'utf-8')
-    return MetricsEvent.fromLogLine(metrics.trim().split('\n').pop())
-  }
-
-  /**
-   * @param {Object} [options]
-   * @param {String} [options.module]
-   * @param {AbortSignal} [options.signal]
-   * @returns {AsyncGenerator<MetricsEvent>}
-   */
-  async * follow ({ module, signal } = {}) {
-    const tail = new Tail(this.#getMetricsFilePath(module), {
-      nLines: 1,
-      useWatchFile: platform() === 'win32'
-    })
-    for await (const [line] of on(tail, 'line', { signal })) {
-      yield MetricsEvent.fromLogLine(line)
-    }
+class Metrics extends EventEmitter {
+  constructor () {
+    super()
+    this.mergedMetrics = null
+    this.moduleMetrics = new Map()
   }
 
-  // Metrics stream
   // - Filters duplicate entries
   // - Writes `jobs-completed` to InfluxDB
-  // - Serializes JSON
-  // - Persists to
-  //   - module specific metrics file
-  //   - merged metrics file
-  async createWriteStream (moduleName) {
-    assert(moduleName, 'moduleName required')
-    const deduplicateStream = new DeduplicateStream()
-    const splitStream = new PassThrough({ objectMode: true })
-    deduplicateStream
-      .pipe(new TelemetryStream(moduleName))
-      .pipe(splitStream)
-    splitStream
-      .pipe(new SerializeStream())
-      .pipe(
-        this.logs.createWriteStream(
-          join(this.metricsPath, `${moduleName}.log`)
-        )
-      )
-    splitStream
-      .pipe(new MergeMetricsStream(...await Promise.all([
-        this.getLatest(),
-        this.getLatest(moduleName)
-      ])))
-      .pipe(new SerializeStream())
-      .pipe(this.logs.createWriteStream(this.allMetricsPath))
-    return deduplicateStream
-  }
-
-  async maybeCreateMetricsFile (moduleName) {
-    await maybeCreateFile(
-      this.#getMetricsFilePath(moduleName),
-      formatLog(
-        JSON.stringify({ totalJobsCompleted: 0, totalEarnings: '0' }) + '\n'
-      )
-    )
-  }
-}
-
-class DeduplicateStream extends Transform {
-  constructor () {
-    super({ objectMode: true })
-    this.last = null
-  }
-
-  _transform (obj, _, callback) {
-    const isChanged = this.last === null ||
-        Object
-          .entries(obj)
-          .some(([key, value]) => this.last[key] !== value)
-    if (isChanged) {
-      this.last = obj
-      callback(null, obj)
-    } else {
-      callback()
-    }
-  }
-}
-
-class TelemetryStream extends Transform {
-  constructor (moduleName) {
-    assert(moduleName, 'moduleName required')
-    super({ objectMode: true })
-    this.lastMetrics = null
-    this.moduleName = moduleName
-  }
-
-  _transform (metrics, _, callback) {
-    if (typeof this.lastMetrics?.totalJobsCompleted === 'number') {
+  // - Merges metrics from all modules
+  submit (moduleName, metrics) {
```

Comment:
What is the type of `metrics`? Should it be `metricsEvent` with type `MetricsEvent`?